File: src/main/java/no/nav/syfo/pdl/PdlConsumer.java
package no.nav.syfo.pdl;

import no.nav.syfo.metric.Metrikk;
import no.nav.syfo.pdl.exceptions.*;
import no.nav.syfo.sts.StsConsumer;
import org.apache.commons.io.IOUtils;
import org.slf4j.Logger;
import org.springframework.beans.factory.annotation.*;
import org.springframework.cache.annotation.Cacheable;
import org.springframework.core.io.ClassPathResource;
import org.springframework.core.io.Resource;
import org.springframework.http.*;
import org.springframework.stereotype.Service;
import org.springframework.web.client.RestClientException;
import org.springframework.web.client.RestTemplate;
import org.springframework.web.util.UriComponentsBuilder;

import java.io.IOException;
import java.io.InputStream;
import java.nio.charset.StandardCharsets;

import static no.nav.syfo.util.RestUtils.bearerHeader;
import static org.slf4j.LoggerFactory.getLogger;
import static org.springframework.http.HttpHeaders.AUTHORIZATION;

@Service
public class PdlConsumer {
    private static final Logger LOG = getLogger(PdlConsumer.class);

    private final Metrikk metrikk;
    private final RestTemplate restTemplate;
    private final StsConsumer stsConsumer;
    private final String url;

    public static final String TEMA_HEADER = "Tema";
    public static final String ALLE_TEMA_HEADERVERDI = "GEN";
    public static final String NAV_CONSUMER_TOKEN_HEADER = "Nav-Consumer-Token";

    @Autowired
    public PdlConsumer(
            Metrikk metrikk,
            @Qualifier("scheduler") RestTemplate restTemplate,
            StsConsumer stsConsumer,
            @Value("${pdl.url}") String url
    ) {
        this.metrikk = metrikk;
        this.restTemplate = restTemplate;
        this.stsConsumer = stsConsumer;
        this.url = url;
    }

    @Cacheable(cacheNames = "personPdl", key = "#ident", condition = "#ident != null")
    public PdlHentPerson person(String ident) {
        metrikk.tellHendelse("call_pdl");

        String query = readFileAsString("pdl/hentPerson.graphql").replace("[\n\r]", "");

        HttpEntity<PdlRequest> entity = createRequestEntity(
                new PdlRequest()
                        .query(query)
                        .variables(new Variables().ident(ident))
        );

        final String uriString = UriComponentsBuilder.fromHttpUrl(url).toUriString();

        try {
            ResponseEntity<PdlPersonResponse> response = restTemplate.exchange(
                    uriString,
                    HttpMethod.POST,
                    entity,
                    PdlPersonResponse.class
            );
            metrikk.tellHendelse("call_pdl_success");
            PdlPersonResponse responseBody = response.getBody();
            throwExceptionIfPDLBodyHasErrorOrIsEmpty(responseBody);
            return responseBody.data;
        } catch (RestClientException e) {
            metrikk.tellHendelse("call_pdl_fail");
            LOG.error("Error from PDL with request-url: " + url, e);
            throw e;
        }
    }

    private void throwExceptionIfPDLBodyHasErrorOrIsEmpty(PdlPersonResponse responseBody) {
        if (responseBody == null) {
            metrikk.tellHendelse("call_pdl_fail");
            throw new EmptyPDLContent("Tried to get name of person :(");
        }
        if (responseBody.errors != null) {
            metrikk.tellHendelse("call_pdl_fail");
            throw new PDLResponseBodyContainsError(responseBody.errors.get(0).toSimplifiedString());
        }
        if (responseBody.data.getName() == null) {
            metrikk.tellHendelse("call_pdl_fail");
            throw new NameFromPDLIsNull("Tried to get name of person :( (SAD) :sadpanda:");
        }
    }

    private HttpEntity<PdlRequest> createRequestEntity(PdlRequest request) {
        String stsToken = stsConsumer.token();

        HttpHeaders headers = new HttpHeaders();
        headers.setContentType(MediaType.APPLICATION_JSON);
        headers.set(TEMA_HEADER, ALLE_TEMA_HEADERVERDI);
        headers.set(AUTHORIZATION, bearerHeader(stsToken));
        headers.set(NAV_CONSUMER_TOKEN_HEADER, bearerHeader(stsToken));

        return new HttpEntity<>(request, headers);
    }

    private String readFileAsString(String path) {
        try {
            Resource resource = new ClassPathResource(path);
            InputStream input = resource.getInputStream();

            return IOUtils.toString(input, StandardCharsets.UTF_8.name());
        } catch (IOException e) {
            metrikk.tellHendelse("pdl_feil");
            LOG.error("Feilet ved lesing av hentPerson query", e);
            throw new RuntimeException("Feilet ved lesing av hentPerson query", e);
        }
    }
}

File: src/main/kotlin/no/nav/syfo/pdl/PdlConsumer.kt
package no.nav.syfo.pdl

import no.nav.syfo.config.CacheConfig.*
import no.nav.syfo.metric.Metrikk
import no.nav.syfo.sts.StsConsumer
import no.nav.syfo.util.*
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.InitializingBean
import org.springframework.beans.factory.annotation.Qualifier
import org.springframework.beans.factory.annotation.Value
import org.springframework.cache.annotation.Cacheable
import org.springframework.core.ParameterizedTypeReference
import org.springframework.http.*
import org.springframework.http.HttpHeaders.AUTHORIZATION
import org.springframework.stereotype.Service
import org.springframework.web.client.RestClientResponseException
import org.springframework.web.client.RestTemplate

@Service
class PdlConsumer(
    private val metric: Metrikk,
    @Value("\${pdl.url}") private val pdlUrl: String,
    private val stsConsumer: StsConsumer,
    @param:Qualifier("scheduler") private val restTemplate: RestTemplate
) : InitializingBean {
    fun person(ident: String): PdlHentPerson? {
        metric.tellHendelse("call_pdl")

        val query = this::class.java.getResource("/pdl/hentPerson.graphql").readText().replace("[\n\r]", "")
        val entity = createRequestEntity(PdlRequest(query, Variables(ident)))
        try {
            val pdlPerson = restTemplate.exchange(
                pdlUrl,
                HttpMethod.POST,
                entity,
                object : ParameterizedTypeReference<PdlPersonResponse>() {}
            )

            val pdlPersonReponse = pdlPerson.body!!
            LOG.info("PDL DATA: ${pdlPersonReponse?.data?.hentPerson?.navn}")
            return if (pdlPersonReponse.errors != null && pdlPersonReponse.errors.isNotEmpty()) {
                metric.tellHendelse("call_pdl_fail")
                pdlPersonReponse.errors.forEach {
                    LOG.error("Error while requesting person from PersonDataLosningen: ${it.errorMessage()}")
                }
                null
            } else {
                metric.tellHendelse("call_pdl_success")
                pdlPersonReponse.data
            }
        } catch (exception: RestClientResponseException) {
            metric.tellHendelse("call_pdl_fail")
            LOG.error("Error from PDL with request-url: $pdlUrl", exception)
            throw exception
        }
    }

    fun personName(ident: String): String? {
        return person(ident)?.fullName()
    }

    fun isKode6Or7(ident: String): Boolean {
        return person(ident)?.isKode6Or7() ?: throw PdlRequestFailedException()
    }

    @Cacheable(cacheNames = [CACHENAME_AKTOER_ID], key = "#fnr")
    fun aktorid(fnr: String): String {
        return hentIdentFraPDL(fnr, IdentType.AKTORID)
    }

    @Cacheable(cacheNames = [CACHENAME_AKTOER_FNR], key = "#aktorId")
    fun fnr(aktorId: String): String {
        return hentIdentFraPDL(aktorId, IdentType.FOLKEREGISTERIDENT)
    }

    @Cacheable(cacheNames = [CACHENAME_GJELDENDE_FNR], key = "#fnr")
    fun gjeldendeFnr(fnr: String): String {
        return hentIdentFraPDL(fnr, IdentType.FOLKEREGISTERIDENT)
    }

    fun hentIdentFraPDL(ident: String, identType: IdentType): String {
        metric.tellHendelse("call_pdl")
        val gruppe = identType.name

        val query = this::class.java.getResource("/pdl/hentIdenter.graphql").readText().replace("[\n\r]", "")
        val entity = createRequestEntity(
            PdlRequest(query, Variables(ident = ident, grupper = gruppe))
        )
        try {
            val pdlIdenter = restTemplate.exchange(
                pdlUrl,
                HttpMethod.POST,
                entity,
                object : ParameterizedTypeReference<PdlIdenterResponse>() {}
            )

            val pdlIdenterReponse = pdlIdenter.body!!
            if (pdlIdenterReponse.errors != null && pdlIdenterReponse.errors.isNotEmpty()) {
                metric.tellHendelse("call_pdl_fail")
                pdlIdenterReponse.errors.forEach {
                    LOG.error("Error while requesting $gruppe from PersonDataLosningen: ${it.errorMessage()}")
                }
                throw RuntimeException("Error while requesting $gruppe from PDL")
            } else {
                metric.tellHendelse("call_pdl_success")
                try {
                    return pdlIdenterReponse.data?.hentIdenter?.identer?.first()?.ident!!
                } catch (e: NoSuchElementException) {
                    LOG.info("Error while requesting $gruppe from PDL. Empty list in hentIdenter response")
                    throw RuntimeException("Error while requesting $gruppe from PDL")
                }
            }
        } catch (exception: RestClientResponseException) {
            metric.tellHendelse("call_pdl_fail")
            LOG.error("Error from PDL with request-url: $pdlUrl", exception)
            throw exception
        }
    }

    private fun createRequestEntity(request: PdlRequest): HttpEntity<PdlRequest> {
        val stsToken: String = stsConsumer.token()
        val headers = HttpHeaders()
        headers.contentType = MediaType.APPLICATION_JSON
        headers.set(PDL_BEHANDLINGSNUMMER_HEADER, BEHANDLINGSNUMMER_OPPFOLGINGSPLAN)
        headers.set(AUTHORIZATION, bearerHeader(stsToken))
        headers.set(NAV_CONSUMER_TOKEN_HEADER, bearerHeader(stsToken))
        return HttpEntity(request, headers)
    }

    companion object {
        private val LOG = LoggerFactory.getLogger(PdlConsumer::class.java)
        lateinit var pdlConsumer: PdlConsumer
    }

    override fun afterPropertiesSet() {
        pdlConsumer = this
    }
}
